### ISSUE TITLE
How does drachtio "recognise" .invalid URIs?

### ISSUE THREAD

**Issue by elbow** [davehorton/drachtio-server#104]
Hi Dave,

I've run into a dead end with my efforts to get a "multi-drachtio" environment to work.

My situation - to handle incoming calls to my mobile clients I must send a push, they then register over WSS and when that happens I release the invite to the contact URI.

The problem being that a drachtio service on drachtio1 is holding the invite, but the client registration was received on drachtio2.

So here's what I did:

1) When drachtio2 receives the register, it is instructed to proxy it to drachtio1.  EG:

Here comes the "push prompted" register:

> 2019-11-11 05:20:24.594628 recv 904 bytes from wss/[10.64.16.102]:40188 at 05:20:24.594388:
REGISTER sip:rtc.xxx.com SIP/2.0
Via: SIP/2.0/WSS 0tji880i7v07.invalid;branch=z9hG4bK1972566
Max-Forwards: 69
To: <sip:2852363@rtc.xxx.com>
From: "2852363" <sip:2852363@rtc.xxx.com>;tag=sh44iebtvj
Call-ID: lldh0adm2fb39g9pecpmo5
CSeq: 1 REGISTER
X-PushId-Platform: android
X-DeviceId: d6911d6bda54d494
X-PushId: dIMeCxic29Y:APA91bHMi_FyJl8HIU81XKklixDIONv7VmD9bIW0khVSZi11_-1p-n1WlsoVxN1L_IugBiutMAqxU52GBSAQaZdz1Pb8kOY53Wi9cqx0Ql99ApnRPKm4YLQetnrzMvRKlWvTGVUKVCzA
X-Rtcproxy-PodId: e2c815dc-0442-11ea-998f-005056
X-Rtcproxy-Auth: auth
Contact: <sip:5uldmu4t@0tji880i7v07.invalid;transport=ws>;+sip.ice;reg-id=1;+sip.instance="<urn:uuid:6b77bb4b-3035-4698-9657-ae3be6c8cd76>";expires=3600
Expires: 3600
Allow: INVITE,ACK,CANCEL,BYE,UPDATE,MESSAGE,OPTIONS,REFER,INFO
Supported: 
User-Agent: xxxxx android 0.43 53
Content-Length: 0

2) We then instruct drachtio to proxy:

> 2019-11-11 05:20:24.597940 http 200 response received from server in 0.00268 secs: {"action":"proxy","data":{"destination":["sip:sip1.xxx.com:6060;transport=tcp"]
,"recordRoute":true}}

Drachtio2 dutifully does the proxy, the register arrives on drachto1 and is processed, the 200 reply is returned on the same path.

(Incidentally - my recordRoute: true is not honoured and there are no Record-Route headers in the proxied REGISTER)

3) My Drachtio service on drachtio one then wants to send the INVITE that it has.  We use Simring.  So it uses a proxy: attribute to get the INVITE proxies back to drachtio2.  We need to do that, by our logic, since that is where the client's Websocket is connected, and that's the only place where "sip:5uldmu4t@0tji880i7v07.invalid" will be recognised.

So:

> [2019-11-11 05:20:24.630 +0000] INFO  (17 on rtcproxy-86975c856b-m8ql2): Starting a simringer with remoteUri
    remoteUri: "sips:5uldmu4t@0tji880i7v07.invalid;transport=ws"
    additionalCallOpts: {
      "headers": {
        "Call-ID": "V1bGRtdTR~rp~594da6de59b9b7a57f4466c1650549a7@xxx.130:5060"
      },
      "proxy": "sip:xxx.157:6060;transport=tcp"
    }

Where "xxx.157" is the IP address of drachtio2

You can see here that I made the URI "sips:" and added ";transport=ws".
I tried is also as the plain "sip:5uldmu4t@0tji880i7v07.invalid" with the same result below

4) So the INVITE sent from drachtio1 arrives on drachtio2:

> INVITE sips:5uldmu4t@0tji880i7v07.invalid;transport=ws SIP/2.0
Via: SIP/2.0/TCP xxx.156:6060;branch=z9hG4bKvvBSB2ty8H8eS
Max-Forwards: 70
From: "Stephen Davies" <sip:2262@xxx.130>;tag=as3966b5f6
To: <sips:5uldmu4t@0tji880i7v07.invalid>;transport=ws
Call-ID: ce59dd8e-7ee5-1238-ab86-00505683aaca
CSeq: 1 INVITE
Contact: <sip:xxx.156:6060;transport=tcp>
In-Reply-To: I4NTIzNjN~rp~r2hduv23u343dquvqluk
Content-Type: application/sdp
Content-Length: 521
X-xxx-Callid: 1573449622.34048189
X-Rtcproxy-CallUUID: 2ca53fb8-9724-48c7-965b-a7f487d2471c

5) We again instruct Drachtio to proxy, this time without a destination which is supposed to result in drachtio-server just looking up where to send it - which is what I want since I don't know.

> 2019-11-11 05:20:24.627329 http 200 response received from server in 0.00188 secs: {"action":"proxy","data":{"recordRoute":true}}

I'd expect it to "recognise" the 5uldmu4t@0tji880i7v07.invalid but seems not, since this is what drachtio2 logs:

> 2019-11-11 05:20:24.627402 DrachtioController::httpCallRoutingComplete thread id 139943985153792 transaction id 1a76de61-8ead-4141-a79e-bd189dea1ebb response: (200) {"action":"proxy","data":{"recordRoute":true}}
2019-11-11 05:20:24.627439 SipProxyController::proxyRequest - transactionId: 1a76de61-8ead-4141-a79e-bd189dea1ebb
2019-11-11 05:20:24.627492 RequestHandler::makeRequestForRoute - after returning handle  in thread139943985153792 4 handles are available in cache
2019-11-11 05:20:24.627518 SipProxyController::doProxy - transactionId: 1a76de61-8ead-4141-a79e-bd189dea1ebb clientMsgId: 
2019-11-11 05:20:24.627567 pending-request: removing entry, prior to removal length: 2
2019-11-11 05:20:24.627602 SipProxyController::addProxy - adding transaction id 1a76de61-8ead-4141-a79e-bd189dea1ebb, id ce59dd8e-7ee5-1238-ab86-00505683aaca|INVITE|1 before insert there are 1 proxy instances
2019-11-11 05:20:24.627637 ProxyCore::initializeTransactions - added 0 client transactions 
2019-11-11 05:20:24.627873 provisional timeout is 5000ms
2019-11-11 05:20:24.627916 ProxyCore::startRequests started 0 client transactions
2019-11-11 05:20:24.627945 Error proxying request; please check that this is a valid SIP Request-URI and retry
2019-11-11 05:20:24.627995 tport.c:4678 tport_by_name() tport(0x55ad1f7c6970): found 0x55ad1f7c5b00 by name TCP/xxx.156:6060
2019-11-11 05:20:24.628023 tport.c:3285 tport_tsend() tport_tsend(0x55ad1f7c5b00) tpn = TCP/197.155.250.156:6060
2019-11-11 05:20:24.628104 tport.c:3531 tport_send_msg() tport_vsend returned 336
2019-11-11 05:20:24.628237 send 336 bytes to tcp/[xxx.156]:6060 at 05:20:24.628056:
SIP/2.0 500 Internal Server Error
Via: SIP/2.0/TCP xxx.156:6060;branch=z9hG4bKvvBSB2ty8H8eS;rport=6060
From: "Stephen Davies" <sip:2262@xxx.130>;tag=as3966b5f6
To: <sips:5uldmu4t@0tji880i7v07.invalid>;transport=ws;tag=6ym6BDc5aU9gN
Call-ID: ce59dd8e-7ee5-1238-ab86-00505683aaca
CSeq: 1 INVITE
Content-Length: 0

I'm obviously missing something in my understanding.  Can you help?

**Comment by davehorton** [comment]
I would recommend doing this differently.

When receiving the REGISTER from a websocket client, do not proxy it or send it anywhere.  Authenticate it locally and send a 200 OK.  That drachtio server will cache the user/host value and associate it with that wss socket. At that point, if that drachtio server gets a request to send an INVITE to that user/host it should lookup that socket and do so.

So then, on the drachtio server receiving the incoming call, send the PUSH, then through some out-of-band message, wait until that client has registered on some drachtio server.  Then b2b the original INVITE over to that server (possibly with some custom headers to indicate what you want to the receiving app to do).  The app that handles incoming invites on the server with the register should just need to b2b the incoming invite to the user/host associated with the register.

This is all probably simpler if you organize drachtio servers into edge-facing servers that terminate wss connections, and "internal" application servers.

Let me know if this does not work

**Comment by davehorton** [comment]
Alternatively, if you want to continue working the approach you outlined, I think the one change you should try making is not to proxy that INVITE on the final leg, but b2b it instead.  The caching of registrations is only used by b2b, not proxy

**Comment by davehorton** [comment]
actually, I could build a test branch that could make the proxy work for you.....?

**Comment by elbow** [comment]
We don't have out of band communication between the rtcproxy instances.

Seems like the problem is that the drachtio has to "see" the register OK to remember the registration? (I hit this before on a single drachtio when I tried to send the invite on to the client BEFORE I OKed the register.) 

And it doesn't "see" the register that is proxied, or the reply? 

Thanks, 
Steve

**Comment by davehorton** [comment]
yes, it has nothing to do ".invalid" actually.  When sending a 200 OK to a REGISTER (but not as proxy), _and_ if the transport is not udp, the drachtio server caches the user/host and associates it with the socket
